File: app/src/main/java/com/elseboot3909/GCRClient/ViewModel/ChangeViewModel.java
package com.elseboot3909.GCRClient.ViewModel;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.elseboot3909.GCRClient.API.ChangesAPI;
import com.elseboot3909.GCRClient.Entities.ChangeInfo;
import com.elseboot3909.GCRClient.Entities.CommitInfo;
import com.elseboot3909.GCRClient.Utils.Constants;
import com.elseboot3909.GCRClient.Utils.JsonUtils;
import com.elseboot3909.GCRClient.Utils.NetManager;
import com.google.gson.Gson;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class ChangeViewModel extends ViewModel {

    private String id;

    private MutableLiveData<ChangeInfo> changeInfo;
    private MutableLiveData<CommitInfo> commitInfo;

    private final Gson gson = new Gson();

    public LiveData<ChangeInfo> getChangeInfo(String id) {
        if (changeInfo == null || !id.equals(this.id)) {
            this.id = id;
            changeInfo = new MutableLiveData<>();
            loadChangeInfo();
        }
        return changeInfo;
    }

    public LiveData<CommitInfo> getCommitInfo(String id) {
        if (commitInfo == null || !id.equals(this.id)) {
            this.id = id;
            commitInfo = new MutableLiveData<>();
            loadCommitInfo();
        }
        return commitInfo;
    }

    private void loadChangeInfo() {
            Retrofit retrofit = NetManager.getRetrofitConfiguration(null, true);

            ChangesAPI changesAPI = retrofit.create(ChangesAPI.class);
            Call<String> retrofitRequest = changesAPI.getChangeDetails(id);

            retrofitRequest.enqueue(new Callback<String>() {
                @Override
                public void onResponse(@NonNull Call<String> call, @NonNull Response<String> response) {
                    if (response.isSuccessful() && response.body() != null) {
                        changeInfo.postValue(gson.fromJson(JsonUtils.TrimJson(response.body()), ChangeInfo.class));
                    } else {
                        Log.e(Constants.LOG_TAG, "onResponse: Not successful");
                    }
                }

                @Override
                public void onFailure(@NonNull Call<String> call, @NonNull Throwable t) {
                    Log.e(Constants.LOG_TAG, "onFailure: Not successful");
                }
            });
    }


    private void loadCommitInfo() {
        Retrofit retrofit = NetManager.getRetrofitConfiguration(null, true);

        ChangesAPI changesAPI = retrofit.create(ChangesAPI.class);
        Call<String> retrofitRequest = changesAPI.getCommitInfo(id, "current");

        retrofitRequest.enqueue(new Callback<String>() {
            @Override
            public void onResponse(@NonNull Call<String> call, @NonNull Response<String> response) {
                if (response.isSuccessful() && response.body() != null) {
                    commitInfo.postValue(gson.fromJson(JsonUtils.TrimJson(response.body()), CommitInfo.class));
                } else {
                    Log.e(Constants.LOG_TAG, "onResponse: Not successful");
                }
            }

            @Override
            public void onFailure(@NonNull Call<String> call, @NonNull Throwable t) {
                Log.e(Constants.LOG_TAG, "onFailure: Not successful");
            }
        });
    }

}

File: app/src/main/java/com/elseboot3909/GCRClient/ViewModel/ChangeViewModel.kt
package com.elseboot3909.GCRClient.ViewModel

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.elseboot3909.GCRClient.API.ChangesAPI
import com.elseboot3909.GCRClient.Entities.ChangeInfo
import com.elseboot3909.GCRClient.Entities.CommitInfo
import com.elseboot3909.GCRClient.Utils.Constants
import com.elseboot3909.GCRClient.Utils.JsonUtils
import com.elseboot3909.GCRClient.Utils.NetManager
import com.google.gson.Gson
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class ChangeViewModel : ViewModel() {

    private var id = ""

    private val changeInfo: MutableLiveData<ChangeInfo> by lazy {
        MutableLiveData<ChangeInfo>().also {
            loadChangeInfo()
        }
    }

    private val commitInfo: MutableLiveData<CommitInfo> by lazy {
        MutableLiveData<CommitInfo>().also {
            loadCommitInfo()
        }
    }

    private val gson = Gson()

    fun getChangeInfo(id: String) : LiveData<ChangeInfo> {
        if (id != this.id) {
            this.id = id
            loadChangeInfo()
        }
        return changeInfo
    }

    fun getCommitInfo(id: String) : LiveData<CommitInfo> {
        if (id != this.id) {
            this.id = id
            loadCommitInfo()
        }
        return commitInfo
    }

    private fun loadChangeInfo() {
            val retrofit = NetManager.getRetrofitConfiguration(null, true)

            val changesAPI = retrofit.create(ChangesAPI::class.java)
            changesAPI.getChangeDetails(id).enqueue(object : Callback<String> {
                override fun onResponse(call: Call<String>, response: Response<String>) {
                    if (response.isSuccessful && response.body() != null) {
                        changeInfo.postValue(gson.fromJson(JsonUtils.trimJson(response.body()), ChangeInfo::class.java))
                    } else {
                        Log.e(Constants.LOG_TAG, "onResponse: Not successful")
                    }
                }

                override fun onFailure(call: Call<String>, t: Throwable) {
                    Log.e(Constants.LOG_TAG, "onFailure: Not successful")
                }
            })
    }


    private fun loadCommitInfo() {
        val retrofit = NetManager.getRetrofitConfiguration(null, true)

        val changesAPI = retrofit.create(ChangesAPI::class.java)
        changesAPI.getCommitInfo(id, "current").enqueue(object : Callback<String> {
            override fun onResponse(call: Call<String>, response: Response<String>) {
                if (response.isSuccessful && response.body() != null) {
                    commitInfo.postValue(gson.fromJson(JsonUtils.trimJson(response.body()), CommitInfo::class.java))
                } else {
                    Log.e(Constants.LOG_TAG, "onResponse: Not successful")
                }
            }

            override fun onFailure(call: Call<String>, t: Throwable) {
                Log.e(Constants.LOG_TAG, "onFailure: Not successful")
            }
        })
    }

}
